Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp

```diff
@@ -179,7 +179,9 @@ void LayerTreeHost::layerFlushTimerFired()
 #endif
 
 #if HAVE(DISPLAY_LINK)
-    m_compositor->updateScene();
+    // Request a composition only if the CompositingCoordinator hasn't called commitSceneState.
+    if (!m_isWaitingForRenderer)
+        m_compositor->updateScene();
```

Comment:
This looks good to me, but to understand it you need to check that `m_isWaitingForRenderer` is set to true in `LayerTreeHost::commitSceneState` that is called from `CompositingCoordinator::flushPendingLayerChanges` only when `shouldSyncFrame` is true. So, I wonder if we can make this easier to follow. I see two options:

* `CompositingCoordinator::flushPendingLayerChanges` returns a boolean that is never read and always returns true. We could return `shouldSyncFrame` and use that value to decide in `LayerTreeHost::layerFlushTimerFired()`
* We could remove `m_compositor->updateScene()` from `LayerTreeHost::layerFlushTimerFired()` and call `m_client.updateScene()` from `CompositingCoordinator::flushPendingLayerChanges` when `shouldSyncFrame` is false

Reply:
> Maybe second option makes more sense, I wonder if it's important to call `updateScene` before `m_page.didUpdateRendering()`, since `commitSceneState()` is called before.

Yes, that makes it easier to understand.
I think `updateScene` can be called before or after `m_page.didUpdateRendering()` without much trouble.